### PR TITLE
NIFI-10048: changed Hashcode function in StandardFlowDifference to ac…

### PR DESCRIPTION
**Pull request by emiliosetiadarma**
…comodate missing identifiers in components, and added tests in TestFlowController that used flow.json files

Still investigating what could cause identifiers to be null. Will follow up in this effort

# Summary

[NIFI-10048](https://issues.apache.org/jira/browse/NIFI-10048)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [x] JDK 11
  - [x] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowComparator.java [hunk 1]
```diff
@@ -129,7 +129,7 @@ private boolean compareComponents(final VersionedComponent componentA, final Ver
     }
 
     private boolean compareComponents(final VersionedComponent componentA, final VersionedComponent componentB, final Set<FlowDifference> differences,
-        final boolean compareName, final boolean comparePos, final boolean compareComments) {
+                                      final boolean compareName, final boolean comparePos, final boolean compareComments) {
```
Comment (exceptionfactory) [inline nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowComparator.java:132]: The changes in this file appear to be limited to formatting, can you revert the changes?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/resources/conf/new-flow.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"encodingVersion":{"majorVersion":2,"minorVersion":0},"maxTimerDrivenThreadCount":10,"registries":[],"parameterContexts":[],"controllerServices":[],"reportingTasks":[],"templates":[],"rootGroup":{"identifier":"2a2b649d-8538-3239-9965-536b5b993cc5","instanceIdentifier":"13c477b8-0182-1000-31df-454d42e70446","name":"NiFi Flow","comments":"","position":{"x":0.0,"y":0.0},"processGroups":[],"remoteProcessGroups":[],"processors":[],"inputPorts":[],"outputPorts":[],"connections":[],"labels":[],"funnels":[],"controllerServices":[],"variables":{},"defaultFlowFileExpiration":"0 sec","defaultBackPressureObjectThreshold":10000,"defaultBackPressureDataSizeThreshold":"1 GB","flowFileOutboundPolicy":"STREAM_WHEN_AVAILABLE","flowFileConcurrency":"UNBOUNDED","componentType":"PROCESS_GROUP"}}
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/resources/conf/new-flow.json:1]: It would be helpful to pretty-print these JSON flow definitions to make them easier to read for testing. Alternatively, is it possible to construct the flow definitions in test methods instead of external files?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/serialization/VersionedFlowSynchronizer.java [hunk 3]
```diff
@@ -384,7 +384,6 @@ private void synchronizeFlow(final FlowController controller, final DataFlow exi
 
     private FlowComparison compareFlows(final DataFlow existingFlow, final DataFlow proposedFlow, final PropertyEncryptor encryptor) {
         final DifferenceDescriptor differenceDescriptor = new StaticDifferenceDescriptor();
-
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/serialization/VersionedFlowSynchronizer.java:387]: This change can be reverted

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the improvements @emiliosetiadarma. The change itself in `StandardFlowDifference.hashCode()` looks good, and the test changes are helpful.

There are several unnecessary formatting changes. If you can revert the formatting changes, that would be helpful.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/resources/conf/new-flow.json (hunk 2 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/resources/conf/new-flow.json:1]: Making the changes to have them set programmatically

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/serialization/VersionedFlowSynchronizer.java (hunk 3 above)
Comment (emiliosetiadarma) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/serialization/VersionedFlowSynchronizer.java:387]: Making the changes

File: nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowComparator.java (hunk 1 above)
Comment (emiliosetiadarma) [reply nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowComparator.java:132]: Will revert the changes

File: nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowComparator.java [hunk 4]
```diff
@@ -624,4 +624,4 @@ private FlowDifference difference(final DifferenceType type, final VersionedComp
     private static interface ComponentComparator<T extends VersionedComponent> {
         void compare(T componentA, T componentB, Set<FlowDifference> differences);
     }
-}
+}
```
Comment (exceptionfactory) [inline nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowComparator.java:627]: Thanks for making the updates @emiliosetiadarma, can you revert this change as well?

File: nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowDifference.java [hunk 5]
```diff
@@ -116,18 +116,18 @@ public boolean equals(final Object obj) {
...
         }
 
         return Objects.equals(componentAId, otherComponentAId) && Objects.equals(componentBId, otherComponentBId)
-            && Objects.equals(description, other.description) && Objects.equals(type, other.type)
-            && Objects.equals(valueA, other.valueA) && Objects.equals(valueB, other.valueB);
+                && Objects.equals(description, other.description) && Objects.equals(type, other.type)
+                && Objects.equals(valueA, other.valueA) && Objects.equals(valueB, other.valueB);
     }
-}
+}
```
Comment (exceptionfactory) [inline nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowDifference.java:133]: This file also contains a number of formatting changes, please revert so that the changes are scoped to the `hashCode()` method. Thanks!

File: nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowDifference.java [hunk 6]
```diff
@@ -89,8 +89,8 @@ public String toString() {
 
     @Override
     public int hashCode() {
-        return 31 + 17 * (componentA == null ? 0 : componentA.getIdentifier().hashCode()) +
-            17 * (componentB == null ? 0 : componentB.getIdentifier().hashCode()) +
+        return 31 + 17 * (componentA == null ? 0 : Objects.hash(componentA.getIdentifier())) +
+            17 * (componentB == null ? 0 : Objects.hash(componentB.getIdentifier())) +
```
Comment (exceptionfactory) [inline nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowDifference.java:93]: This method should use Objects.`hashCode(Object o)` instead of Objects.`hash(Object... values)`. The `Objects.hash()` method delegates to `Arrays.hashCode()`, but all that is necessary is to use `Objects.hashCode()` to provide null-safe handling of `identifier`.
```suggestion
        return 31 + 17 * (componentA == null ? 0 : Objects.hashCode(componentA.getIdentifier())) +
            17 * (componentB == null ? 0 : Objects.hashCode(componentB.getIdentifier())) +
```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for reverting the formatting changes @emiliosetiadarma. The new tests look good.

On closer inspection of `StandardFlowDifference.hashCode()`, it looks like `Objects.hashCode()` should be used instead of `Objects.hash()`.

File: nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowDifference.java (hunk 6 above)
Comment (emiliosetiadarma) [reply nifi-registry/nifi-registry-core/nifi-registry-flow-diff/src/main/java/org/apache/nifi/registry/flow/diff/StandardFlowDifference.java:93]: Hey David, thanks for the catch, will make those changes

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the adjustments @emiliosetiadarma, the latest version looks good. +1 merging
